File: backend/app/schemas/map.py
"""Map payloads; mirrored by electron/ui/src/types/map.ts.

Coordinates use longitude/latitude order. KML retains string coordinates and
altitude; JSON from shapefiles may contain numbers. Neither representation is
normalized here, and arbitrary source attributes remain attached to each record.
"""
from pydantic import Field

from .base import PayloadModel

CoordinateValue = int | float | str
MapCoordinates = list[CoordinateValue]


class MapFacilityFields(PayloadModel):
    """Facility form values mirrored by MapFacilityFields in TypeScript."""
    TreatmentTechnology: str | None = None
    Capacity: CoordinateValue | None = None
    PadWaterQuality: CoordinateValue | None = None
    PadStorageInitialWaterQuality: CoordinateValue | None = None
    StorageInitialWaterQuality: CoordinateValue | None = None
    ExternalWaterQuality: CoordinateValue | None = None
    InitialDisposalCapacity: CoordinateValue | None = None
    InitialStorageCapacity: CoordinateValue | None = None
    CompletionsPadStorage: CoordinateValue | None = None
    PadOffloadingCapacity: CoordinateValue | None = None
    NodeCapacities: CoordinateValue | None = None
    DisposalOperationalCost: CoordinateValue | None = None
    ReuseOperationalCost: CoordinateValue | None = None
    ExternalSourcingCost: CoordinateValue | None = None
    TruckingHourlyCost: CoordinateValue | None = None
    DesalinationSites: CoordinateValue | None = None
    BeneficialReuseCost: CoordinateValue | None = None
    BeneficialReuseCredit: CoordinateValue | None = None
    CompletionsPadOutsideSystem: CoordinateValue | None = None
    Elevation: CoordinateValue | None = None
    SWDDeep: CoordinateValue | None = None
    SWDAveragePressure: CoordinateValue | None = None
    SWDProxPAWell: CoordinateValue | None = None
    SWDProxInactiveWell: CoordinateValue | None = None
    SWDProxEQ: CoordinateValue | None = None
    SWDProxFault: CoordinateValue | None = None
    SWDProxHpOrLpWell: CoordinateValue | None = None


class MapNode(MapFacilityFields):
    node_type: str
    name: str | None = None
    nodeType: str | None = None
    coordinates: MapCoordinates | None = None
    longitude: CoordinateValue | None = None
    latitude: CoordinateValue | None = None
    altitude: CoordinateValue | None = None


class ArcNodeRef(PayloadModel):
    name: str
    incoming: bool | None = None
    outgoing: bool | None = None
    coordinates: MapCoordinates | None = None
    incoming_nodes: list[str] | None = None
    outgoing_nodes: list[str] | None = None
    # Geometry between this node and the next; flow comes from outgoing_nodes.
    segment_coordinates: list[MapCoordinates] | None = None


class MapArc(PayloadModel):
    name: str | None = None
    node_type: str | None = None
    coordinates: list[MapCoordinates] | None = None
    nodes: list[ArcNodeRef] | None = None
    lengths: list[int | float] | None = None
    # Older map editors persisted the aggregate length as text.
    length: CoordinateValue | None = None
    diameter: CoordinateValue | None = None


class ConnectionMetadata(PayloadModel):
    pipeline_capacity: CoordinateValue | None = None
    pipeline_length: CoordinateValue | None = None
    pipeline_diameter: CoordinateValue | None = None


class MapConnections(PayloadModel):
    all_connections: dict[str, list[str]] = Field(default_factory=dict)
    connection_metadata: dict[str, ConnectionMetadata] | None = None


class MapData(PayloadModel):
    all_nodes: dict[str, MapNode] = Field(default_factory=dict)
    arcs: dict[str, MapArc] = Field(default_factory=dict)
    connections: MapConnections = Field(default_factory=MapConnections)
    ProductionPads: dict[str, MapNode] = Field(default_factory=dict)
    CompletionsPads: dict[str, MapNode] = Field(default_factory=dict)
    NetworkNodes: dict[str, MapNode] = Field(default_factory=dict)
    SWDSites: dict[str, MapNode] = Field(default_factory=dict)
    TreatmentSites: dict[str, MapNode] = Field(default_factory=dict)
    StorageSites: dict[str, MapNode] = Field(default_factory=dict)
    ExternalWaterSources: dict[str, MapNode] = Field(default_factory=dict)
    ReuseOptions: dict[str, MapNode] = Field(default_factory=dict)
    other_nodes: dict[str, MapNode] = Field(default_factory=dict)
    # Map form keys use decision_period; workbook units use "decision period".
    units: dict[str, str] | None = None
    defaultNode: str | None = None
    time_periods: list[str] | None = None
    node_renames: dict[str, str] | None = Field(default=None, alias='_node_renames')

File: electron/ui/src/types/map.ts
import type React from 'react';
import type {Scenario} from './scenario';

/** Map units object */
export interface MapUnits {
  volume: string;
  distance: string;
  diameter: string;
  concentration: string;
  currency: string;
  time: string;
  pressure: string;
  elevation: string;
  decision_period: string;
  mass: string;

  [k: string]: string;
}

/** Coordinates stay in longitude/latitude order until converted for Leaflet.
 * KML retains strings and optional altitude; shapefiles and editor drafts may use numbers.
 */
export type CoordinateValue = number | string;
export type MapFieldValue = number | string | null;
export type MapCoordinates = CoordinateValue[];

/** Facility form values, including blanks/nulls before scenario completion. */
export interface MapFacilityFields {
  TreatmentTechnology?: string;
  Capacity?: MapFieldValue;
  PadWaterQuality?: MapFieldValue;
  PadStorageInitialWaterQuality?: MapFieldValue;
  StorageInitialWaterQuality?: MapFieldValue;
  ExternalWaterQuality?: MapFieldValue;
  InitialDisposalCapacity?: MapFieldValue;
  InitialStorageCapacity?: MapFieldValue;
  CompletionsPadStorage?: MapFieldValue;
  PadOffloadingCapacity?: MapFieldValue;
  NodeCapacities?: MapFieldValue;
  DisposalOperationalCost?: MapFieldValue;
  ReuseOperationalCost?: MapFieldValue;
  ExternalSourcingCost?: MapFieldValue;
  TruckingHourlyCost?: MapFieldValue;
  DesalinationSites?: MapFieldValue;
  BeneficialReuseCost?: MapFieldValue;
  BeneficialReuseCredit?: MapFieldValue;
  CompletionsPadOutsideSystem?: MapFieldValue;
  Elevation?: MapFieldValue;
  SWDDeep?: MapFieldValue;
  SWDAveragePressure?: MapFieldValue;
  SWDProxPAWell?: MapFieldValue;
  SWDProxInactiveWell?: MapFieldValue;
  SWDProxEQ?: MapFieldValue;
  SWDProxFault?: MapFieldValue;
  SWDProxHpOrLpWell?: MapFieldValue;
}

export interface MapNode extends MapFacilityFields {
  name?: string;
  longitude?: CoordinateValue;
  latitude?: CoordinateValue;
  altitude?: CoordinateValue;
  coordinates?: MapCoordinates;
  node_type: string;
  nodeType?: string;
  /** Preserve unrecognized source metadata; consumers must narrow it before use. */
  [key: string]: unknown;
}

/** A reference to a facility along a pipeline, not a separate facility record. */
export interface ArcNodeRef {
  name: string;
  incoming?: boolean;
  outgoing?: boolean;
  coordinates?: MapCoordinates;
  incoming_nodes?: string[];
  outgoing_nodes?: string[];
  /** Bends from this node to the next geometric node, retained across unchanged segments. */
  segment_coordinates?: MapCoordinates[];
  [key: string]: unknown;
}

export interface MapArc {
  name?: string;
  node_type?: string;
  coordinates?: MapCoordinates[];
  nodes?: ArcNodeRef[];
  lengths?: number[];
  length?: number;
  diameter?: CoordinateValue;
  [key: string]: unknown;
}

export type ConnectionsMap = Record<string, string[]>;

export interface ConnectionMetadata {
  pipeline_capacity?: MapFieldValue;
  pipeline_length?: MapFieldValue;
  pipeline_diameter?: MapFieldValue;
  [key: string]: unknown;
}

export interface MapConnections {
  all_connections: ConnectionsMap;
  /** Directed edge keys use "from::to". */
  connection_metadata?: Record<string, ConnectionMetadata>;
  [key: string]: unknown;
}

/** Imported maps may omit category views until workbook preprocessing rebuilds them. */
export interface MapData {
  all_nodes: Record<string, MapNode>;

  ProductionPads?: Record<string, MapNode>;
  CompletionsPads?: Record<string, MapNode>;
  NetworkNodes?: Record<string, MapNode>;
  SWDSites?: Record<string, MapNode>;
  TreatmentSites?: Record<string, MapNode>;
  StorageSites?: Record<string, MapNode>;
  ExternalWaterSources?: Record<string, MapNode>;
  ReuseOptions?: Record<string, MapNode>;
  other_nodes?: Record<string, MapNode>;

  connections: MapConnections;
  arcs: Record<string, MapArc>;

  units?: MapUnits;
  defaultNode?: string;
  time_periods?: string[];
  /** Transient rename instructions consumed by the backend before persistence. */
  _node_renames?: Record<string, string>;

  [k: string]: unknown;
}

// Map / Editor related types
export type CoordinateTuple = [number | string, number | string];

/** Editor records reuse facility values but distinguish UI kind from facility type. */
export interface MapEditorNode extends MapFacilityFields {
  name?: string;
  node_type?: string;
  nodeType?: string;
  coordinates?: MapCoordinates;
  lengths?: number[];
  nodes?: ArcNodeRef[];
  length?: number;
  diameter?: CoordinateValue;
  [key: string]: unknown;
}

/** The provider starts with an empty array before the first map is loaded. */
export type MapContextData = Partial<MapData> | ([] & {defaultNode?: never; units?: never});

export type SelectedNodeState = {
  node: MapEditorNode;
  idx: number;
};

export interface MapContextValue {
  networkMapData: MapContextData;
  setNetworkMapData: React.Dispatch<React.SetStateAction<MapContextData>>;
  selectedNode: SelectedNodeState | null;
  setSelectedNode: React.Dispatch<React.SetStateAction<SelectedNodeState | null>>;
  showNetworkNode: boolean;
  setShowNetworkNode: React.Dispatch<React.SetStateAction<boolean>>;
  showNetworkPipeline: boolean;
  setShowNetworkPipeline: React.Dispatch<React.SetStateAction<boolean>>;
  addNode: () => void;
  addPipeline: () => void;
  clickNode: (node: MapEditorNode, idx: number) => void;
  clickPipeline: (node: MapEditorNode, idx: number) => void;
  saveNodeChanges: (updatedNode: MapEditorNode, deselectAfterwards?: boolean) => void;
  nodeData: MapEditorNode[];
  setNodeData: React.Dispatch<React.SetStateAction<MapEditorNode[]>>;
  lineData: MapEditorNode[];
  setLineData: React.Dispatch<React.SetStateAction<MapEditorNode[]>>;
  handleMapClick: (coords: CoordinateTuple) => void;
  availableNodes: MapEditorNode[];
  creatingNewNode: boolean;
  selectingPipelineConnectionFromMap: boolean;
  setSelectingPipelineConnectionFromMap: React.Dispatch<React.SetStateAction<boolean>>;
  pipelineConnectionSelectionIndex: number | null;
  setPipelineConnectionSelectionIndex: React.Dispatch<React.SetStateAction<number | null>>;
  deleteSelectedNode: () => void;
  currentlyCreatingPipeline: boolean;
  currentlyCreatingNode: boolean;
  handleFileUpload: (file: File, defaultNodeType?: string) => void;
  nodeType: "node" | "pipeline" | null;
}

export interface MapProviderProps {
  children: React.ReactNode;
  scenario?: Scenario | null;
  handleUpdateScenario: (updatedScenario: Scenario | null, setScenarioData?: boolean, propagateChanges?: string) => void;
}

export interface MapAdditionalField {
  key: keyof MapFacilityFields;
  displayName: string;
  type: 'number' | 'boolean' | 'dropdown' | 'dynamic_dropdown';
  defaultValue?: number | string | null;
  defaultOptions?: Array<number | string>;
  booleanValues?: {true: string; false: string};
  reliesOn?: keyof MapFacilityFields;
  usesKey?: 'TreatmentCapacityIncrements';
  units?: string;
  unit?: string;
  tip?: string;
}
